feat: add semantic helper functions choose_file/choose_folder/choose_tmp and allow all semantic helper functions to be used when parsing resources

fgvieira opened the pull request:
<!--Add a description of your PR here-->
Sometimes, HPC nodes have different options to write `temp` depending on the hardware (e.g. some might have `nvme`).

This function allows the use to specify a list of possible temp folders that, on each job submission, is evaluated. The first valid `dir` is selected.

Moved `dynamic` resources to after `standard` and `default` resources.

### QC
<!-- Make sure that you can tick the boxes below. -->

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Adds a helper to select a usable temporary directory and exposes it for use in workflows and resource expressions.
  * Resource expressions can now reference ioutils helpers during evaluation.

* **Tests**
  * Workflow tests consolidated to exercise dynamic tmpdir selection via profile-based runs; one legacy default-only test removed.

* **Documentation**
  * Expanded guidance and examples for standard/default resources, dynamic resources, and tmpdir configuration.

* **Chores**
  * CI test matrix parallelism increased for faster runs.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

fgvieira committed:
Add get_tmp function

fgvieira committed:
Code refactor

fgvieira committed:
Code format

fgvieira committed:
Update test

fgvieira committed:
Add docs

fgvieira committed:
Fix typo

fgvieira committed:
Add example with env variable

fgvieira committed:
Add slurm jobid example to docs

fgvieira committed:
Rename function

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Fix docs

fgvieira committed:
Fix doc typos

fgvieira committed:
Merge branch 'main' into tmp_list

johanneskoester commented:
Why not just relying on the tmpdir resource, which can be configured by the admins or the user via the standard tmpdir env variable?

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira commented:
The issue is when HPC have different hardware configurations. In my case, some have a fast `nvme` scratch space, while others only have a slow `nfs` scratch. Since we never know where the jobs will be submitted to, this decision has to be made at runtime.

The best solution would definitely be to have the admins define the `best` temp folder into a env variable, and use it in tmpdir resource. However, it might not be possible.
As for the user defining the tempdir via the standard tmpdir env variable, how would you do it at runtime?

The idea here is to define a helper function that can be passed to `resources.tmpdir` to select the first valid folder of a list or, if none, `system_tmpdir`. For example:
```
resources:
    tmpdir = choose_tmp(["/fast_nvme", "/home/$USER/scratch", "/scratch/$SLURM_JOB_ID"])
```
This function would be executed at runtime and, depending on the node, it would select the first folder that meets certain criteria to be a tmpdir.

johanneskoester commented on src/snakemake/ioutils/tmp.py:
```diff
@@ -0,0 +1,33 @@
+from pathlib import Path
+
+
+def choose_tmp(tmpdirs: List) -> str:
```
```suggestion
def choose_tmp(tmpdirs: List[Union[Path, str]]) -> str:
```

johanneskoester commented on src/snakemake/ioutils/__init__.py:
```diff
@@ -25,5 +26,6 @@ def register_in_globals(_globals):
             "parse_input": parse_input,
             "extract_checksum": extract_checksum,
             "flatten": flatten,
+            "choose_tmp": choose_tmp,
```
I think in addition this needs to be present in the eval environment for resource expressions, right?

johanneskoester replied:
As `ioutils` though, right? Maybe rather the individual functions? This way, it would be symmetric between the snakefile and resource definitions.

fgvieira committed:
Update src/snakemake/ioutils/tmp.py

Co-authored-by: Johannes Köster <johannes.koester@uni-due.de>

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Update src/snakemake/ioutils/tmp.py

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

fgvieira committed:
Add profile example

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Fix typo

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

fgvieira committed:
Fix typo

fgvieira committed:
Merge tests

fgvieira committed:
Fix tests

cademirch committed:
Merge branch 'main' into tmp_list

cademirch commented:
This is a nice addition @fgvieira - are you wanting to wait until the hackathon to get this merged? If its okay I'll see if I can get the tests passing and merge this sooner.

cademirch committed:
ci: set max-parallel for test matrix.

Pixi install is failing with "failed to parse pypi name mapping" errors
likely due to rate limiting

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Update main.yml

fgvieira committed:
Fix typo

fgvieira committed:
Revert doc changes

fgvieira committed:
Update docs

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Code refactor

fgvieira committed:
Typo

fgvieira committed:
Try to accept Path

fgvieira committed:
Code format

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Add check for file/folder

fgvieira committed:
Fix tests

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Add wildcards

fgvieira committed:
Merge branch 'main' into tmp_list

fgvieira committed:
Code format

johanneskoester committed:
Merge branch 'main' into tmp_list

johanneskoester commented:
@fgvieira what do you think about the generalization towards arbitrary dirs?

fgvieira commented:
@johanneskoester from what I remember, I think that I have already implemented that with the additions (apart from `choose_tmp`) of two extra generic functions (`choose_file` and `choose_dir`).

Or are you thinking about something else?

fgvieira committed:
Merge branch 'main' into tmp_list

johanneskoester commented:
Ah, right, sorry, I now remember that I had already seen that.

johanneskoester commented on docs/snakefiles/rules.rst:
```diff
@@ -833,17 +833,52 @@ The ``subpath`` function can be very handy in combination with :ref:`Snakemake's
 
+.. _snakefiles-choose-f:
+
+The choose_f(ile/older) functions
```
I think this section could need some better explanation. (a) the parameters of `choose_f*`, namely read/write/etc. (b) the assertions in the examples can be confusing to non-python-experts. I would rather avoid the asserts and instead explain what the respective calls of choose_file/folder return (e.g. in python comments.

fgvieira committed:
Update docs

fgvieira committed:
Update choose_f.py

fgvieira committed:
fix typo

johanneskoester committed:
simplify code

johanneskoester committed:
Update resources.py

johanneskoester committed:
Refactor namespace update for better readability
